tools/generate_code_cache.js

```diff
@@ -0,0 +1,97 @@
+'use strict';
+
+const natives = process.binding('natives');
+const vm = require('vm');
+const fs = require('fs');
+let count = 0;
+
+function human(num) {
+  if (num < 1024) {
+    return `${(num).toFixed(2)}B`;
+  } else if (num < 1024 ** 2) {
+    return `${(num / 1024).toFixed(2)}KB`;
+  } else if (num < 1024 ** 3) {
+    return `${(num / (1024 ** 2)).toFixed(2)}MB`;
+  } else {
+    return `${(num / (1024 ** 3)).toFixed(2)}GB`;
+  }
+}
+
+/**
+ * @param {string} key
+ * @param {Buffer} cache
+ */
+function getInitalizer(key, cache) {
+  const defName = key.replace(/\//g, '_').replace(/-/g, '_');
+  const def = `static uint8_t ${defName}_raw[] = {\n` +
+                     `${cache.join(',')}\n};`;
+  const initializer = `
+  v8::Local<v8::ArrayBuffer> ${defName}_ab =
+    v8::ArrayBuffer::New(isolate, ${defName}_raw, ${cache.length});
+  v8::Local<v8::Uint8Array> ${defName}_array =
+    v8::Uint8Array::New(${defName}_ab, 0, ${cache.length});
+  target->Set(context,
+              OneByteString(isolate, "${key}"),
+              ${defName}_array).FromJust();
+  `;
+  return {
+    def, initializer
+  };
+}
+
+const cacheDefs = [];
+const initializers = [];
+
+for (const key of Object.keys(natives)) {
+  if (key === 'config') continue;
+  const wrapper = [
+    '(function (exports, require, module, process) {',
+    '\n});'
+  ];
+  const code = wrapper[0] + natives[key] + wrapper[1];
+
+  const script = new vm.Script(code, {
+    filename: `${key}.js`,
+    produceCachedData: true
+  });
+
+  if (script.cachedData) {
+    const length = script.cachedData.length;
+    count += length;
+    const { def, initializer } = getInitalizer(key, script.cachedData);
+    cacheDefs.push(def);
+    initializers.push(initializer);
+    console.log(`Added ${key} = ${human(length)}` +
+                `, total = ${human(count)}`);
+  } else {
+    const { def, initializer } = getInitalizer(key, Buffer.alloc(0));
+    cacheDefs.push(def);
+    initializers.push(initializer);
```

Comment:
Is there a reason why we are still adding the definition and the initializer to the output source file when the cached data failed to generate?

Reply:
@TimothyGu I think it makes sense to make the build step more lenient. If the cache data cannot be created it'll just be empty and in the current implementation it's going to be added to `cacheRejectedList`. If we are going to throw on bad caches we can skip the cache if it's an empty buffer.

Reply:
Ah I see. Thanks for the explanation.